platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java

```diff
@@ -293,10 +295,39 @@ public ConfigurableFileCollection unset() {
     public void setFrom(Object... paths) {
         assertMutable();
         setExplicitCollector(paths.length > 0
-            ? newValue(value, paths)
+            ? newExplicitValue(paths)
             : EMPTY_COLLECTOR);
     }
 
+    private ValueCollector newConventionValue(Iterable<?> paths) {
+        return newValue(EMPTY_COLLECTOR, paths);
+    }
+
+    private ValueCollector newConventionValue(Object[] paths) {
+        return newValue(EMPTY_COLLECTOR, paths);
+    }
+
+    private ValueCollector newExplicitValue(Iterable<?> paths) {
+        return newValue(getBaseValue(), paths);
+    }
+
+    private ValueCollector newExplicitValue(Object[] paths) {
+        return newValue(getBaseValue(), paths);
+    }
+
+    private ValueCollector getBaseValue() {
+        if (!isExplicit() && !value.isEmpty()) {
+            return copySources(value);
+        }
+        return value;
+    }
```

Comment:
:question: Correct me if I'm wrong, but my understanding of this is "if there is no explicit value, then return a copy of the convention's collector; otherwise return the current value". However, all callsites are in `setFrom`, which discards the copied paths anyway. This looks suboptimal.

Reply:
I will hve to revisit this later - but has to do with the fact `setFrom()` used to (and still) mutates the current value collector. That was fine before conventions. Now we need to be careful not to reuse the value collector from a convention when assigning an explicit value, as that breaks the invariant that convention values are seperate from explicit values. The code we have in place avoids such sharing for conventions. It may have other caveats though.

Reply:
@mlopatkin Please see #27663, which means to address both explicit values leaked into conventions, and vice-versa.